File: Java/GenericStructures/ForwardList.java
package GenericStructures;

public class ForwardList<T> {
    class Node{
        T data;
        Node next = null;
        Node(T data){
            this.data = data;
        }
    }

    private Node front = null;
    private Node back = null;
    private int size;

    public ForwardList(){
        this.size = 0;
    }

    public void add(T data){
        Node node = new Node(data);

        if(this.is_empty()){
            this.front = node;
            this.back = node;
        }
        else{
            this.back.next = node;
            back = node;
        }
        this.size += 1;
    }

    public void add_front(T data){
        Node node = new Node(data);

        if(this.is_empty()){
            this.front = node;
            this.back = node;
        }
        else{
            node.next = this.front;
            this.front = node;
        }
        this.size += 1;
    }

    public void add_back(T data){
        Node node = new Node(data);

        if(this.is_empty()){
            this.front = node;
            this.back = node;
        }
        else{
            this.back.next = node;
            back = node;
        }
        this.size += 1;
    }

    public void remove(T data){
        // Node node = new Node(data);
        try {
            boolean is_element = false;
            Node ptr = this.front;
            Node prev = ptr;
            while(ptr.next != null){
                if(ptr.data == data){
                    is_element = true;
                    prev.next = ptr.next;
                    this.size -= 1;
                    return;
                }
                prev = ptr;
                ptr = ptr.next;
            }

            if(is_element == false) throw new Exception();

        } catch (Exception e) {
            System.out.print(String.valueOf(data) + " is not an element of the list.");
        }
    }

    public void remove_first(){
        Node node = this.front;
        this.front = node.next;
        node = null;

        this.size -= 1;
    }

    public void remove_last(){
        Node ptr = this.front;
        while(ptr.next != null)
            ptr = ptr.next;

        ptr = null;

        this.size -= 1;
    }

    public void display(){
        Node ptr = this.front;
        
        while(ptr.next != null){
            System.out.print(String.valueOf(ptr.data) + " ");
            ptr = ptr.next;
        }
    }

    boolean is_empty(){
        return this.front == null;
    }

    public int size(){return this.size(); }
}

class Main{
    public static void main(String [] args){
        ForwardList<Integer> list = new ForwardList<>();
        for(int i = 0; i<5; ++i)
            list.add(i);
        
        list.display();
    }
}

File: C#/GenericStructures/ForwardList.cs
using System;

class ForwardList <T> where T : IComparable<T>{
    public class Node{
        public T data;
        public Node next = null;
        public Node(T data){
            this.data  = data;
        }
    }

    Node front = null;
    Node back = null;

    int m_size;

    public ForwardList(){
        this.m_size = 0;
    }

    public void add(T data){
        Node node  = new Node(data);
        if(is_empty()){
            this.front = node;
            this.back = node;
        }
        else{
            this.back.next = node;
            this.back = node;
        }

        ++ this.m_size;
    }

    public void add_front(T data){
        Node node = new Node(data);
        if (is_empty()){
            this.front = node;
            this.back = node;
        }else{
            node.next = this.front;
            this.front = node;            
        }


        this.m_size += 1;
    }

    public void add_back(T data){
        Node node = new Node(data);
        if (is_empty()){
            this.front = node;
            this.back = node;
        }else{
            node.next = this.back;
            this.back = node;
        }

        this.m_size += 1;
    }

    public void remove(T data){
        try{
            // if(is_empty()) throw System.Exception;

            if(data.CompareTo(this.front.data) == 0) remove_front();

            Node ptr = this.front;
            Node prev = ptr;
            
            while(ptr.next != null){
                if(ptr.data.CompareTo(data) == 0){
                    if(ptr.data.CompareTo(this.back.data) == 0) {
                        remove_back(prev);
                        return;
                    }

                    prev.next = ptr.next;
                    return;
                }
                prev = ptr;
                ptr = ptr.next;
            }
            -- this.m_size;

        }catch (System.Exception){
            throw; // handle exception
        }
    }

    private void remove_front(){
        if(is_empty()) return; // improve with exception handling
        Node temp = this.front;
        this.front = this.front.next; 
        temp = null;
        -- this.m_size;
    }

    private void remove_back(Node prev){
        if(is_empty()) return; // improve with exception handling
        Node temp = this.back;
        this.back = prev;
        temp = null;

        -- this.m_size;
    }
    
    public void display(){
        Node ptr = this.front;
        while(ptr.next != null){
            Console.Write(ptr.data + " ");
            ptr = ptr.next;
        }
    }
    
    public bool is_empty(){return this.front == null;}

    public int size(){return this.m_size;}
}

class MainList{
    public static void Main(string [] args){
        ForwardList list = new ForwardList();

        for(int i = 0; i<10; ++i)
            list.add(i);
            
        list.remove(2);
        list.display();
    }
}